src/org/labkey/test/tests/InvalidateSessionTest.java

```diff
@@ -0,0 +1,164 @@
+            response = selectCmd.execute(cn, getProjectName());
+            Assert.assertEquals("Did not establish the database connection before the password change", 200,
+                    response.getStatusCode());
+        }
+        catch (IOException | CommandException e)
+        {
+            throw new RuntimeException(e);
+        }
+
+        log("Changing the user password");
+        String newPassword = PasswordUtil.getPassword() + "&*&*";
+        goToChangePassword().setOldPassword(PasswordUtil.getPassword())
+                .setPassword1(newPassword)
+                .setPassword2(newPassword)
+                .clickSubmit();
+
+        log("Verifying the session is invalid");
+        try
+        {
+            selectCmd.execute(cn, getProjectName());
+            fail("Test should have thrown command exception");
+        }
+        catch (CommandException e)
+        {
+            Assert.assertEquals("Session should be expired", 401, e.getStatusCode());
+        }
+        signOut();
+    }
+
+    /*
+        Regression coverage for Secure Issue 31493: Test for session and cookie persistence through login and logout
+     */
+    @Test
+    public void testCookieAndSessionFromLogout() throws IOException
+    {
+        goToProjectHome();
+
+        log("Capture the cookie after login");
+        Set<Cookie> beforeCookie = getDriver().manage().getCookies();
+
+        log("Establish the connection");
+        Connection cn = createDefaultConnection();
+        SelectRowsResponse response;
+        SelectRowsCommand selectCmd = new SelectRowsCommand("auditLog", "UserAuditEvent");
+        try
+        {
+            response = selectCmd.execute(cn, getProjectName());
+            Assert.assertEquals("Did not establish the database connection before the password change", 200,
+                    response.getStatusCode());
+        }
+        catch (CommandException e)
+        {
+            throw new RuntimeException(e);
+        }
+
+        log("Sign out");
+        signOut();
+
+        log("Verify previously established connection be used after log out");
+        try
+        {
+            selectCmd.execute(cn, getProjectName());
+            fail("Should fail when logged out");
+        }
+        catch (CommandException e)
+        {
+            Assert.assertEquals("Command execution should fail when logged out", HttpStatus.SC_UNAUTHORIZED, e.getStatusCode());
+        }
+
+        log("Capture the cookie after logout");
+        Set<Cookie> afterCookie = getDriver().manage().getCookies();
+        Assert.assertFalse("Before and after log out cookie should be different", beforeCookie.equals(afterCookie));
+    }
+
+    private SetPasswordForm goToChangePassword()
+    {
+        if (PasswordUtil.getUsername().equals(getCurrentUser()))
+            throw new IllegalArgumentException("Don't change the primary site admin user's password");
+
+        goToMyAccount();
+        clickButton("Change Password");
```

Comment:
Add the "Change Password" button to the `UserDetailsPage` returned by `goToMyAccount`. Also, move the primary user assertion into the method that clicks it.
Also, update `PasswordTest` where this appears to be copied from.

Reply:
Done